cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/TimePartitionedFileSetDefinition.java

```diff
@@ -0,0 +1,109 @@
+import co.cask.cdap.api.dataset.DatasetSpecification;
+import co.cask.cdap.api.dataset.lib.AbstractDatasetDefinition;
+import co.cask.cdap.api.dataset.lib.CompositeDatasetAdmin;
+import co.cask.cdap.api.dataset.lib.FileSet;
+import co.cask.cdap.api.dataset.lib.FileSetArguments;
+import co.cask.cdap.api.dataset.lib.TimePartitionedFileSet;
+import co.cask.cdap.api.dataset.lib.TimePartitionedFileSetArguments;
+import co.cask.cdap.api.dataset.table.Table;
+import co.cask.cdap.explore.client.ExploreFacade;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Maps;
+import com.google.inject.Injector;
+import com.google.inject.Provider;
+
+import java.io.IOException;
+import java.util.Map;
+import javax.inject.Inject;
+
+/**
+ * Defines the partitioned dataset type. At this time, the partitions are not managed by the
+ * partitioned dataset, so all admin is simply on the partition table.
+ */
+public class TimePartitionedFileSetDefinition extends AbstractDatasetDefinition<TimePartitionedFileSet, DatasetAdmin> {
+
+  private static final String PARTITION_TABLE_NAME = "partitions";
+  private static final String FILESET_NAME = "files";
+
+  private final DatasetDefinition<? extends Table, ?> tableDef;
+  private final DatasetDefinition<? extends FileSet, ?> filesetDef;
+
+  @Inject
+  private Injector injector;
+
+  public TimePartitionedFileSetDefinition(String name,
+                                          DatasetDefinition<? extends FileSet, ?> filesetDef,
+                                          DatasetDefinition<? extends Table, ?> tableDef) {
+    super(name);
+    Preconditions.checkArgument(tableDef != null, "Table definition is required");
+    Preconditions.checkArgument(filesetDef != null, "FileSet definition is required");
+    this.filesetDef = filesetDef;
+    this.tableDef = tableDef;
+  }
+
+  @Override
+  public DatasetSpecification configure(String instanceName, DatasetProperties properties) {
+    return DatasetSpecification.builder(instanceName, getName())
+      .properties(properties.getProperties())
+      .datasets(filesetDef.configure(FILESET_NAME, properties),
+                tableDef.configure(PARTITION_TABLE_NAME, properties))
+      .build();
+  }
+
+  @Override
+  public DatasetAdmin getAdmin(DatasetSpecification spec, ClassLoader classLoader) throws IOException {
+    return new CompositeDatasetAdmin(filesetDef.getAdmin(spec.getSpecification(FILESET_NAME), classLoader),
+                                     tableDef.getAdmin(spec.getSpecification(PARTITION_TABLE_NAME), classLoader));
+  }
+
+  @Override
+  public TimePartitionedFileSet getDataset(DatasetSpecification spec,
+                                           Map<String, String> arguments, ClassLoader classLoader)
+    throws IOException {
+    // if the arguments do not contain an output location, generate one from the partition time
+    if (FileSetArguments.getOutputPath(arguments) == null) {
```

Comment:
For inputs, we can't do it in the getDataset(), because it requires reading the partition table, and I don't think we should do data operations inside getDataset() - there is no transaction here. 

For inputs, we do that lookup at the time that we generate the arguments for the input format. We could do the same for the output path, but then the file set itself would not have know the output path, and the MR job would not be able to get that in the onFinish(). 

Once I address the TODO that registration of the partition should be done in the output committer of the output format (which means I have to write an output format wrapper that loads the dataset, delegates record writing to he dataset's output format, but then defines its own output committer), then we don't need this path in the onFinish() any longer, and I can move this logic. I would like to do that in a separate PR.